<a href="https://colab.research.google.com/github/lmmlima/ENV716_EnergyModeling_F2021/blob/main/Lab2/Lab2_Exploring_egrid_SOLUTION.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Lab 2 – Reading Data Sets in Python**

Learning goals for Lab 2:
1. How to read excel spreadsheets in Python;
2. How to slice you large data set into a smaller data set;
3. How to filter rows and columns;
4. How to get summary values.


If you have not mounted google drive yet please do so by running the code chunk below. 

In [1]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Set up work directory

When reading/importing a file in Python notebook you need to inform the path where the file is located. If you simply specify the name of the file, Python will assume it is in your working directory. If the file is not there, the code will output a “file not found” error message. 

The fisrt step is to find out your where is your working directory. If you don’t know, import the os library and use the command os.getcwd() to find out.


In [2]:
import os
os.getcwd()

'/content'

In the function os.getcwd(), getcwd stands for get current working directory. Note that my working directory is “/content/”. If I want to use read_excel() and just specify the name of my excel file. I need to have a copy of the file in the folder “/content/”.

Ideally you should create a Folder inside your google drive to store the Python scripts we are creating during the labs. That way you will not have all files in the same place. For example, I have a folder insive my drive called "Colab Notebooks".



Note that e-grid file and all .ipynb notebooks we created so far are in the "ColabNotebooks" folder.
And the path to the folder is:

/content/drive/MyDrive/ColabNotebooks/


If you want to change your working directory to your "ColabNotebooks"  folder you can use function os.chdir() - chdir stands for change directory. And inside the brackets you type the path to the folder. In my computer it would be as below.

In [3]:
os.chdir('/content/drive/MyDrive/ColabNotebooks/')

You can check with working directory changed by using again the os.getcwd().

In [4]:
os.getcwd()

'/content/drive/MyDrive/ColabNotebooks'

Note that now when you use the os.getcwd() command, the new working directory is not “/content/” anymore, instead it’s pointing to my the "ColabNotebooks" folder. Once you do that, you are all set and ready to import the e-grid file.


## Importing Excel file

We will use the function read_excel() from **pandas library** to import the file. Remember we do not need to install any libraries with Colab, but we do need to import them using the code chunck below. A good practice in Python coding is to rename the library with a short name that will be used anytime you need to refer to a function from that library. 

For example if you ant to use the reac_excel from pandas you code will be 
**pandas.read_excel()**
If you rename pandas as pd then you can simply use
**pd.read_excel()**

In [5]:
import pandas as pd
import numpy as np

The function **read_excel()** takes two main arguments:
*   io: file path and name;
*   sheet_name: if your workbook has more than one spreadsheet, you need to identify which spreadsheet you want to read.

Both arguments take string type values that’s why we are using the quotes. When you use quotes, Colab will automatically change the text to red to indicate it's a string. The other arguments for the read_excel() are optional ones. 

When you import a spreadsheet, you need to store it in your Pyhton environment as a variables. A common data structure used for spreadsheet is the data frame because is allows columns of different type (numerical or string) to be sotored in the same variable. The pd.read_excel() function will import you spreadshete as a data frame and store in a variable of your choice. Here I will call is **df_gen**. The df is my way of remembering it's a data frame and the gen tell me it's generators information. 

A good practive when coding is to always assign informative names to your variables.

The **.head()** function is used to inspect the data frame. Any number inside the parenthesis will be the number of row to be printed.



In [6]:
df_gen = pd.read_excel(io="egrid2019_data.xlsx",sheet_name="GEN19")
df_gen.head(5)

eGRID2019 Generator file sequence number  ... Generator planned or actual retirement year
0                                 SEQGEN19  ...                                    GENYRRET
1                                        1  ...                                         NaN
2                                        2  ...                                         NaN
3                                        3  ...                                         NaN
4                                        4  ...                                         NaN

[5 rows x 17 columns]

Note that because egrid had 2 rows explaining the data, the second row is mistakenly added to the table, to fix this you can add an argument inside de read_excel() functions called **skiprows**. As the name suggests skiprows is used when you need to skiprows in your spreadsheet before importing the data set.


In [7]:
df_gen = pd.read_excel(io="egrid2019_data.xlsx",sheet_name="GEN19",skiprows=1)
df_gen.head(5)

SEQGEN19  YEAR PSTATABB  ... GENERSRC  GENYRONL GENYRRET
0         1  2019       AK  ...      NaN       NaN      NaN
1         2  2019       AK  ...      NaN    1977.0      NaN
2         3  2019       AK  ...      NaN    1977.0      NaN
3         4  2019       AK  ...      NaN    1977.0      NaN
4         5  2019       AK  ...      NaN    1977.0      NaN

[5 rows x 17 columns]

Note that now our header is exactly the 2 second of our worksheet. And the first row was not imported.  

We still have a problem here. The NaN. NaN stands for Not a Number. E-grid has a lot of empty cell and Python reads them as NaN.
Depending on the type of data you have, you may want to replace with something else. 

You can specify with the read_excel function what you want to do with missing values using the argument **keep_default_na = False** as in the code chunk below. 


In [8]:
df_gen = pd.read_excel(io="egrid2019_data.xlsx",sheet_name="GEN19",skiprows=1,keep_default_na=False)
df_gen.head(5)

SEQGEN19  YEAR PSTATABB  ... GENERSRC  GENYRONL GENYRRET
0         1  2019       AK  ...                            
1         2  2019       AK  ...               1977         
2         3  2019       AK  ...               1977         
3         4  2019       AK  ...               1977         
4         5  2019       AK  ...               1977         

[5 rows x 17 columns]

## Manipulating the data frame (data set)

Now your the data frame is ready to be used and manipualted. Imagine you want to select only a subset of the columns. You can start by  inspecting columns names by using the code below.

In [9]:
df_gen.columns

Index(['SEQGEN19', 'YEAR', 'PSTATABB', 'PNAME', 'ORISPL', 'GENID', 'NUMBLR',
       'GENSTAT', 'PRMVR', 'FUELG1', 'NAMEPCAP', 'CFACT', 'GENNTAN', 'GENNTOZ',
       'GENERSRC', 'GENYRONL', 'GENYRRET'],
      dtype='object')

Now you can use use function .loc() to select columns of interest. Choose a name for you new subset data frame. Here we are calling it **sub_df_gen**. And say that you want to select state, name, code, fuel, nameplate capacity and capacity factor of the generator. 

In [10]:
sub_df_gen = df_gen.loc[:,['PSTATABB','PNAME','ORISPL','FUELG1','NAMEPCAP','CFACT']]
sub_df_gen.head(5)


PSTATABB                             PNAME  ORISPL FUELG1 NAMEPCAP CFACT
0       AK         7-Mile Ridge Wind Project   60814    WND      1.8      
1       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5      
2       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5      
3       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5      
4       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5

If you want to use the column index number instead of the column name you need the function .iloc instead of .loc. From our original data framde df_gen we know the state is on the third columns, but Python starts counting from 0, so you need to point to column 2 to get PSTATABB, 3 to get PNAME and 4 to get ORISPL as shown in the code chunk below.

Note the notation here df_gen is a data frame so we refer to columns and row using the brackets [], to point to all elements we use : and a subset of element we use 2:5. Altough we are only looking for columns 2,3,4 we need to specify 2:5. 

In [11]:
sub_df_gen = df_gen.iloc[:,2:5]
sub_df_gen.head(5)

PSTATABB                             PNAME  ORISPL
0       AK         7-Mile Ridge Wind Project   60814
1       AK  Agrium Kenai Nitrogen Operations   54452
2       AK  Agrium Kenai Nitrogen Operations   54452
3       AK  Agrium Kenai Nitrogen Operations   54452
4       AK  Agrium Kenai Nitrogen Operations   54452

What about the other columns? If you want to specify more then one range for columns you need the np.r_[] function. Recall no stands for numpy.  

In [12]:
sub_df_gen = df_gen.iloc[:,np.r_[2:5,9:12]]
sub_df_gen.head(5) 

PSTATABB                             PNAME  ORISPL FUELG1 NAMEPCAP CFACT
0       AK         7-Mile Ridge Wind Project   60814    WND      1.8      
1       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5      
2       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5      
3       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5      
4       AK  Agrium Kenai Nitrogen Operations   54452     NG      2.5

Some can be done for a subset of rows. Say you want rows 101 to 109, then you specify the range 101:110.

In [13]:
sub_df_gen = df_gen.iloc[101:110,]
sub_df_gen.head(10)

SEQGEN19  YEAR  ... GENYRONL GENYRRET
101       102  2019  ...     2015         
102       103  2019  ...     1983         
103       104  2019  ...     1989         
104       105  2019  ...              2015
105       106  2019  ...     1976         
106       107  2019  ...     2009         
107       108  2019  ...     2013         
108       109  2019  ...     2008         
109       110  2019  ...     1988         

[9 rows x 17 columns]

## Filtering the data

But what if you want a specific subset if rows, say you are interested in **wind generators only**. You can **filter** your data by only selecting the rows whose columns FUELG1 is equal to "WND". Remember that WND stands for wind power.

Note here I am accessing the column in a different fashion. So instead of writing df_gen[“FUELG1”]=="WND" I am using sub_df_gen.FUELG1=="WND". Both are correct!


In [14]:
sub_df_gen = df_gen.iloc[:,np.r_[2:5,9:12]]  ##going back to subset data frame with the columns of insterest
wnd_df_gen = sub_df_gen[(sub_df_gen.FUELG1=="WND")]
wnd_df_gen

PSTATABB                               PNAME  ...  NAMEPCAP  CFACT
0           AK           7-Mile Ridge Wind Project  ...       1.8       
106         AK                     Delta Wind Farm  ...       0.9  0.239
107         AK                     Delta Wind Farm  ...       0.9  0.239
108         AK                     Delta Wind Farm  ...       0.1  0.239
171         AK                      Eva Creek Wind  ...      24.6  0.279
...        ...                                 ...  ...       ...    ...
29121       WY                     Seven Mile Hill  ...      24.1  0.301
29126       WY               Silver Sage Windpower  ...        42  0.326
29146       WY                            TB Flats  ...     503.2       
29147       WY  Top of the World Windpower Project  ...       200  0.313
29148       WY              TransAlta Wyoming Wind  ...       144  0.272

[1686 rows x 6 columns]

Note that all rows have "WND" in the FUELG1 column. Now what if I want wind and solar power only. I can add another condition inside the brackets.

In [15]:
wnd_sun_df_gen = sub_df_gen[(sub_df_gen.FUELG1=="WND")|(sub_df_gen.FUELG1=="SUN")]
wnd_sun_df_gen

PSTATABB                               PNAME  ...  NAMEPCAP  CFACT
0           AK           7-Mile Ridge Wind Project  ...       1.8       
106         AK                     Delta Wind Farm  ...       0.9  0.239
107         AK                     Delta Wind Farm  ...       0.9  0.239
108         AK                     Delta Wind Farm  ...       0.1  0.239
171         AK                      Eva Creek Wind  ...      24.6  0.279
...        ...                                 ...  ...       ...    ...
29126       WY               Silver Sage Windpower  ...        42  0.326
29140       WY                    Sweetwater Solar  ...        92  0.224
29146       WY                            TB Flats  ...     503.2       
29147       WY  Top of the World Windpower Project  ...       200  0.313
29148       WY              TransAlta Wyoming Wind  ...       144  0.272

[6838 rows x 6 columns]

## Summarizing the data

Now imagine that you want to sum nameplate capacity to see total installed capacity for wind and sun. You can start with our subset data wnd_sun_df_gen that already contain our generators of interest. All you have left to do is to point to the columns you want to sum and use function .sum(). 

In [16]:
sum_cap=wnd_sun_df_gen['NAMEPCAP'].sum()
sum_cap

230310.99999999953

Remember the units for NAMEPCAP column in e-grid is MW. Therefore, the total installed capacity for sun and wind in 2019 was 230,310 MW or 230 GW.

## Exercise 1

What if you want separated values for wind and sun? 

Part a) Write a code that will give you the total installed capacity in 2019 for **wind** power.

In [17]:
## your code
wnd_df_gen = sub_df_gen[(sub_df_gen.FUELG1=="WND")]
sum_cap_wnd = wnd_df_gen['NAMEPCAP'].sum()

Part b) Write a code that will give you the total installed capacity in 2019 for ***solar*** power.

In [18]:
## your code
sun_df_gen = sub_df_gen[(sub_df_gen.FUELG1=="SUN")]
sum_cap_sun = sun_df_gen['NAMEPCAP'].sum()

Part c) Now sum the total installed capcity from wind (part 1) and solar (part 2) and see if you got the same 230 GW.

In [19]:
## your code
sum_cap_wnd + sum_cap_sun

230311.00000000023

## Exercise 2 - from lecture handout H3

1.	Look at Hatfield’s Ferry Power Station. Where is located? How many units? How many generators? Capacity factor in 2019? What happened to this plant?


In [20]:
## your code
## Need to import UNIT19 spreadsheet to find out number of units and generators
df_unit = pd.read_excel(io="egrid2019_data.xlsx",sheet_name="UNT19",skiprows=1,keep_default_na=False)
df_unit.head(10)
## Name to filter should match name in the spreadsheet
df_unit[(df_unit.PNAME == "Hatfields Ferry Power Station")]

##Also filtering df_gen to find out capacity factor
df_gen[(df_gen.PNAME == "Hatfields Ferry Power Station")]

SEQGEN19  YEAR PSTATABB  ... GENERSRC  GENYRONL GENYRRET
22854     22855  2019       PA  ...                        2013
22855     22856  2019       PA  ...                        2013
22856     22857  2019       PA  ...                        2013

[3 rows x 17 columns]

**Answer** Hatfields Ferry Power Station is located in Pennsylvania. It has three units with 1 generator each. The capacity factor in 2019 is zero or blank. The units retired in 2013.

2.	Are there many Power Plants that have a blank for Capacity Factor in 2019? 
How many Units have blank ?



In [21]:
## your code
gen_blank_cfact = df_gen[(df_gen.CFACT == "")]
## function len() gives you the number of element in a list/vector or number of rows in data frame
len(gen_blank_cfact)

6087

**Answer** 6087 units have blank capacity factor.

3. How many units have been retired?


In [22]:
## your code
#gen_retired = df_gen[(df_gen.GENYRRET <= 2019)]

df_gen.dtypes



SEQGEN19     int64
YEAR         int64
PSTATABB    object
PNAME       object
ORISPL       int64
GENID       object
NUMBLR       int64
GENSTAT     object
PRMVR       object
FUELG1      object
NAMEPCAP    object
CFACT       object
GENNTAN     object
GENNTOZ     object
GENERSRC    object
GENYRONL    object
GENYRRET    object
dtype: object

In [23]:
df_gen['GENYRRET']=pd.to_numeric(df_gen['GENYRRET'],downcast='signed')

df_gen.dtypes

df_gen.head()


SEQGEN19  YEAR PSTATABB  ... GENERSRC  GENYRONL GENYRRET
0         1  2019       AK  ...                         NaN
1         2  2019       AK  ...               1977      NaN
2         3  2019       AK  ...               1977      NaN
3         4  2019       AK  ...               1977      NaN
4         5  2019       AK  ...               1977      NaN

[5 rows x 17 columns]

In [24]:
gen_retired = df_gen[(df_gen.GENYRRET <= 2019)]
len(gen_retired)

3792

**Answer** 3792 units have been retired.

4. How many retired units that use BIT as primary fuel?



In [25]:
## your code
gen_retired_BIT = gen_retired[(gen_retired.FUELG1 == "BIT")]
len(gen_retired_BIT)

447

5. What is the average age of all the units in the database?



In [26]:
## your code
df_gen['GENYRONL']=pd.to_numeric(df_gen['GENYRONL'],downcast='signed')
mu = df_gen['GENYRONL'].mean()
age = 2021 - mu
age

30.040085679314643

6. What is the average age of all units that burn bituminous “BIT” coal?



In [27]:
## your code
df_gen_BIT = df_gen[(df_gen.FUELG1 == "BIT")]
mu = df_gen_BIT['GENYRONL'].mean()
age = 2021 - mu
age

47.65467625899282


7. What is the average age of all units without SO2 scrubber and burning Bituminous coal?



In [28]:
## your code
# Need to look at UNT19 tab
unit_BIT_noScrubber = df_unit[(df_unit.FUELU1 == "BIT") & (df_unit.SO2CTLDV == "")]
len(unit_BIT_noScrubber)

295

8. What is the average age of all units built after 1990? 



In [29]:
## your code
df_unit['UNTYRONL']=pd.to_numeric(df_unit['UNTYRONL'],downcast='signed')

df_unit_1990 = df_unit[(df_unit.UNTYRONL >= 1990)]
mu = df_unit_1990['UNTYRONL'].mean()
age = 2021 - mu
age

17.938744832769544

9. What is the average year when all plants in the database came online? Or find the average age.

In [30]:
## your code
mu = df_unit['UNTYRONL'].mean()
age = 2021 - mu
age


29.733000475511062